### ISSUE TITLE
Multi-language support determined by the browser

### ISSUE THREAD

**Issue by agross** [ioBroker/ioBroker.lovelace#344]
**Is your feature request related to a problem? Please describe.**

I'd like to manage my ioBroker backend in English, but the browser's preference should determine the language used by Lovelace.

**Describe the solution you'd like**

I should be able to define `smartName`s for my entities and Lovelace should display the matching `smartName` or fall back to the ioBroker backend language.

```json
{
  "type": "device",
  "common": {
    "name": "Kitchen Table Light",
    "smartName": {
      "en": "Kitchen Table Light",
      "de": "Küchenlicht Tisch"
    }
  },
  "_id": "zigbee.0.b4e3f9fffe683239",
  // ...
}
```

When the browser requests the Lovelace UI with e.g. `Accept-Language: de-DE,de;q=0.9,en-US,en;q=0.8` Lovelace should display `Küchenlicht Tisch`, otherwise `Kitchen Table Light`.

**Describe alternatives you've considered**

Having everything in English.

**Additional context**

As of today the friendly name (display name) for an entity is determined by the server and [is pinned to the ioBroker backend language](https://github.com/ioBroker/ioBroker.lovelace/blob/acc4ab4c4a5812af9e1f0a1e7743bb5e6644decd/lib/server.js#L893).

**Comment by agross** [comment]
Hello again,

after trying to understand and playing with the code I realize that there need to be some changes made to implement my use-case. Before I start coding please let me know whether you would be OK with them.

My goal is to provide guests with a German Lovelace UI that is the same setup as the one I use in English. So there will be 2 Lovelace adapter instances. I have scripts in place to update the German views/card config when I make changes to the English config. I do *not* want to enter my entity friendly/display names manually, but rather get from ioBroker states. 

There are auto-detected entities and manually-created ones. For manual entities, I can use this config to have the display name translated depending on the Lovelace instance. This works:

```json
{
  "type": "state",
  "common": {
    "name": "Is open",
    "custom": {
      "lovelace.0": {
        "enabled": true,
        "entity": "binary_sensor",
        "name": "Entrance_Door_Contact",
        "attr_device_class": "opening",
        "attr_friendly_name": "Entrance Door"
      },
      "lovelace.1": {
        "enabled": true,
        "entity": "binary_sensor",
        "name": "Entrance_Door_Contact",
        "attr_device_class": "opening",
        "attr_friendly_name": "Eingang Tür"
      }
    }
  },
  "_id": "zigbee.0.00158d0007104fae.opened"
  // ...
}
```

Currently, there is no such support for auto-detected entities (e.g. Hue lights) because the display name is read from the device object itself, and devices cannot have `common.custom`. 

Using `common.smartName` might work, but as of today ioBroker.lovelace's translation always picks the ioBroker Admin language, as stated above.

I'm OK with dropping browser language detection, so the client-side JavaScript can stay as-is. For entities created on the server I suggest the following:

* Add a language setting to ioBroker.lovelace, defaulting to the ioBroker Admin language if unset
* Determine **entity IDs** always based on the English `smartName` (if there are multiple languages available) such that the view/card config is language-independent
* Pick the **friendly/display name** from `smartName` for the chosen ioBroker.lovelace  language setting above
* `common.name` serves as a fallback as before

To summarize, the priorities for entity IDs will be (higher to lower):
1. `common.smartName.en`
1. `common.smartName` if `string`
1. `common.name`
1. The current logic to determine the ID from function and room

The priorities for entity friendly/display names will be:
1. `common.custom.lovelace.<number>.attr_friendly_name`
1. `common.smartName.<ioBroker.lovelace instance language>` if instance language is set
1. `common.smartName.<ioBroker Admin language>`
1. `common.smartName.en` (this could be discussed, but matches other frameworks that I use)
1. `common.smartName` if `string`
1. `common.name`
1. The current logic to determine the name

Please let me know what you think.

**Comment by Garfonso** [comment]
Why would you want to change entity_id to smartName? Currently, smartName is not used for entity_id, so this would be a breaking change (because in existing configurations the entity_ids might change, which can require huge changes). I don't really see why you would need that for your use case.

The entity_id for automatically detected devices is generated here:
https://github.com/ioBroker/ioBroker.lovelace/blob/master/lib/converters/utils.js#L404

So it will always use the English part of common.name.

I'm not against setting a desired frontend language, which then determines the friendly names.

**Comment by Garfonso** [comment]
Why should smartName be more stable than name? Most objects don't even have a smartName set, AFAIK. smartName is used by iot (i.e. alexa voice control).

**Comment by agross** [comment]
> Why should smartName be more stable than name?

I don't understand. Smart Name supports per-language-settings. This is what I want to use instead of using what the admin interface dictates w.r.t. to the display name.

> smartName is used by iot (i.e. alexa voice control).

Perhaps. I use Smart Names quite a lot.

**Comment by Garfonso** [comment]
Ok, let's restart. It seems we are misunderstanding each other.

I'm fine with your proposal. Only this part is not ok:

> To summarize, the priorities for entity IDs will be (higher to lower):
> 
>     1. `common.smartName.en`
> 
>     2. `common.smartName` if `string`
> 
>     3. `common.name`
> 
>     4. The current logic to determine the ID from function and room

You add `common.smartName.en` /  `common.smartName` to the entity_id logic for no good reason -> so I'd say, just leave entity_id as is. It does not make sense to change entity_id handling for language support in the UI, because entity_id is not shown there.

**Comment by agross** [comment]
Yes, I got that wrong in my initial proposal. Let's define:

- entity id: the thing you add in your cards in design mode, e. g. light.Bedroom. It needs to remain stable.
- display name or friendly name: the thing that users see on the UI. For this I want to change/enhance the rules as per your quote of my proposal in the last reply

**Comment by Garfonso** [comment]
Ok. Great. :-)

Feel free to write me if you need any help with your PR.

**Comment by agross** [comment]
@Garfonso

Another question came up:

Some existing tests assert the entity ID and friendly name assigned to manually created entities in a way that doesn't make sense to me. For example, [this tests](https://github.com/ioBroker/ioBroker.lovelace/blob/acc4ab4c4a5812af9e1f0a1e7743bb5e6644decd/lib/converters/utils.test.js#L100-L119) asserts that `entity.entity_id ==  "light.Kitchen"` and `entity.attributes.friendly_name == "Manual defined name"`. I think the code is mixing up the concept of "name". (IMHO, it should always talk about entity IDs or friendly names, just "name" confuses me always despite looking at the code for the whole day.)

The value of `Manual defined name` would not be accepted by the admin UI when you set that value using the "entity name" field (== entity ID, I guess): `Entity name contains invalid characters, allowed are: a-zA-Z0-9А-Яа-я_`. From my understanding the value should never be used as a friendly name. Who would want a friendly name with underscores in it?

Let's assume the value of `state.common.custom.lovelace.<number>.name` should be used for entity IDs only, not friendly names. This would break some of the existing tests:

```mocha
  1) converters/utils
       processCommon -> manual name:

      AssertionError: expected { entity_id: 'light.Kitchen', …(2) } to have nested property 'attributes.friendly_name' of 'Manual defined name', but got 'test.0.airCondition'
      + expected - actual

      -test.0.airCondition
      +Manual defined name

      at expectEntity (lib/converters/utils.test.js:11:35)
      at Context.<anonymous> (lib/converters/utils.test.js:519:9)
      at processImmediate (internal/timers.js:464:21)

  2) converters/utils
       processCommon -> unit:

      AssertionError: expected { …(3) } to have nested property 'attributes.friendly_name' of 'Manual defined name', but got 'test.0.airCondition.unit'
      + expected - actual

      -test.0.airCondition.unit
      +Manual defined name

      at expectEntity (lib/converters/utils.test.js:11:35)
      at Context.<anonymous> (lib/converters/utils.test.js:541:9)
      at processImmediate (internal/timers.js:464:21)
```

It would also break one of the new tests that I added. It seems we need to make the distinction what the sources of entity ID and friendly name are more clear and get rid of the too generic "name" concept.

I would appreciate your input.

**Comment by Garfonso** [comment]
Hm.. yeah.. there are some confusions at some parts in there... it is partly, because a lot of code is from Bluefox and I probably did not unterstand all of it directly. Also there are much more parts, than you see here:

1. entity_id
2. entity name seem to be light.XYZ -> the XYZ part in HomeAssistant. 
3. friendly_name (which is what the lovelace UI will show as name of the device)
4. common.name in ioBroker Objects

> The value of `Manual defined name` would not be accepted by the admin UI when you set that value using the "entity name" field (== entity ID, I guess): `Entity name contains invalid characters, allowed are: a-zA-Z0-9А-Яа-я_`. From my understanding the value should never be used as a friendly name. Who would want a friendly name with underscores in it?

The UI you talk about (the manually configured entity, right?) is asking for 2, not for 3. 

But in both cases (manual and automatic entity generation) the friendly_name is derived from the common.name (i.e. 4) of the ioBroker object. 

> Let's assume the value of `state.common.custom.lovelace.<number>.name` should be used for entity IDs only, not friendly names. This would break some of the existing tests:

This is already true today. And the tests won't break for that. To me, it seems more like the tests don't like the common.name field, which is the sole source for the friendly_name attribute today.

**Comment by agross** [comment]
> The UI you talk about (the manually configured entity, right?) is asking for 2, not for 3.

Yes, the LL-related admin parts where you set name, device_class etc. It's asking for the name (e.g. `foo`) which will be part of the `entity_id` (e.g. `sensor.foo`)

>  The value of `Manual defined name` would not be accepted by the admin UI

Did you see this part? The test uses data that is would not be accepted.

> But in both cases (manual and automatic entity generation) the friendly_name

> This is already true today. 

Yes, but the failing tests assert it is derived from the instance-specific entity name!

> To me, it seems more like the tests don't like the common.name field,

There is no `common.name` for both `obj`s used by the tests.
